### PR TITLE
Use Serializable for Tokenizers

### PR DESCRIPTION
**Pull request by odashi**
This pull request implements `Serializable` interface to `Tokenizer`s and removes existing code that directly treats dicts.

This PR also fixes some typing/interface issues around `tokenizer.py` encountered during applying this change, and moved code in `integration_tests/tokenizers_test.py` to `explainaboard/utils/tokenizer_test.py` because it is actually unit tests.

### CODE REVIEW

File: explainaboard/utils/tokenizer.py [hunk 1]
```diff
@@ -68,55 +120,51 @@ def detokenize(self, tokens: list[str]) -> str:
...
 class SingleSpaceTokenizer(Tokenizer):
     """
     Split a string on a single ascii space
     """
 
     @lru_cache(maxsize=20)
     def __call__(self, text: str) -> TokenSeq:
-        ret = TokenSeq()
         start = 0
+        strs: list[str] = []
```
Comment (neubig) [inline explainaboard/utils/tokenizer.py:155]: Just out of curiosity, I was wondering why this change is made. Is it because the dataclass was turned into `final`? (no changes to the code are necessary)

**Review by neubig — APPROVED** [review]
Looks great, thanks!

File: explainaboard/utils/tokenizer.py (hunk 1 above)
Comment (odashi) [reply explainaboard/utils/tokenizer.py:155]: It is because of separation of concerns. This function requires no information from Tokenizer because the arguments provide sufficient information to construct TokenSeq.

Comment (odashi) [reply explainaboard/utils/tokenizer.py:155]: Maybe it was better to separate this change from this pr to keep the whole change simple.

Comment (neubig) [reply explainaboard/utils/tokenizer.py:155]: Oh, sorry, my comment was not very clear. I meant: "no changes to this PR are necessary, I'm just asking out of curiosity".

Comment (odashi) [reply explainaboard/utils/tokenizer.py:155]: Nice, thanks for the clarification!
